### PR TITLE
feat(Metrics): Combine table- and plot metrics

### LINKED ISSUE #189: Add new accuracy metric for machine learning applications
In machine learning applications we very often care about the accuracy of the trained model (rather than/together with optimality). The idea then is to add a new metric that computes the accuracy of a trained model.

The following should be accomplished:

- [ ] add a `forward` method to data-driven costs, to allow evaluation of the trained model on a set of data
- [ ] customize the way accuracy is displayed in tables (as percentage not scientific notation); new issue will be opened

### LINKED ISSUE #220: Refactor: Combine PlotMetrics and TableMetrics into a Single Metrics Class
### Proposal
Refactor the codebase to unify `PlotMetrics` and `TableMetrics` into a single `Metrics` class. The aim is to simplify the metric interface and usage for benchmarks.

#### Main points:
- Remove the current `PlotMetrics` and `TableMetrics` classes entirely.
- Introduce a new `Metrics` superclass.
    - `Metrics` should implement `get_plot_data()` and `get_table_data()`.
    - The API should encourage users to only implement a single method `get_data(i)` which returns the metric data for iteration `i`.
        - `get_table_data()` becomes just: `get_data(-1)`
        - `get_plot_data()` becomes: `[(i, get_data(i)) for i in iters]`
- Usage: To keep the existing interface, assign metric objects to either the `plot_metrics` or `table_metrics` lists within a benchmark function; the `Metrics` objects can be flexibly assigned to either as needed.
- This should greatly simplify the interface and reduce duplication.

#### Benefits:
- Less boilerplate for users when adding new metrics
- Consistent API for both plotting and table metrics
- Easier maintenance, more extensible metric design

#### Next Steps:
- Agree on the design and naming conventions.
- Refactor codebase to implement the unified `Metrics` class and remove old metric classes.
- Update documentation and usages across benchmarks.

### LINKED ISSUE #241: Fix create_regression_problem
Linear regression is used for regression, logistic regression for classificaiton. Currently `create_regression_problem` creates a classification problem, and the user can choose to apply linear regression cost to it (which doesn't make sense). We should have two problem generation utilities, one for regression and one for classification (which is the current one)

### PR DESCRIPTION
**Pull request by Simpag**
This PR combines table- and plot metrics into one class to make it easier to create new metrics. Users are only required to assign table- and plot descriptions and implement the abstract method 
```python
def get_data_from_trial(
    self, 
    agents: Sequence[AgentMetricsView], 
    problem: BenchmarkProblem, 
    iteration: int,
) -> list[float]:
```
which is used to retrieve metric data at iteration `iteration`, if `iteration` is `-1` then the final iteration should be used. Furthermore, if more than 3 plot metrics are to be plotted, they are either split up into groups of 3 and plotted as subplots in their own figures or uses a user specified grouping by providing plot metrics to the `benchmark` function as `list[list[Metric]]` where each sub-list is a figure. A few more empirical-cost oriented metrics are also added:
* Accuracy
* MSE
* Precision
* Recall

This PR also includes a new cost function for `PyTorch` neural network models. This cost function is optimized to be used within decent bench, the major optimizations are:
* a much more efficient per-gradient calculation than the naive approach
* efficient gradient flattening (to vectors)
* data loading schemes

Finally this PR fixes `create_regression_problem` by only accepting `LinearRegressionCost | PyTorchCost` and adding a `create_classification_problem` with a new `SyntheticClassificationDatasetHandler` which accepts `LogisticRegressionCost | PyTorchCost`

closes #220 closes #189 closes #241

### CODE REVIEW

File: decent_bench/costs/_empirical_risk/_pytorch_cost.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+            return
+
+        try:
+            self._ft_compute_sample_grad = torch.compile(self._ft_compute_sample_grad, **self._compile_kwargs)
+        except Exception as e:
+            LOGGER.warning(f"Error compiling per-sample gradient function: {e}\n\nContinuing without compilation.")
+
+    def __add__(self, other: Cost) -> Cost:
+        if self.shape != other.shape:
+            raise ValueError(f"Mismatching domain shapes: {self.shape} vs {other.shape}")
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_empirical_risk/_pytorch_cost.py]: I think its fine

File: decent_bench/costs/_empirical_risk/_linear_regression_cost.py [hunk 2]
```diff
@@ -215,11 +230,28 @@ def gradient(self, x: NDArray[float64], indices: EmpiricalRiskIndices = "batch")
...
         return res
 
+    def _per_sample_gradients(
+        self,
+        x: NDArray[float64],
+        indices: EmpiricalRiskIndices = "batch",
+    ) -> NDArray[float64]:
+        A, ATA, b = self._get_batch_data(indices)  # noqa: N806
+        res = [ATA[i, :].reshape(-1, 1) * x - A[i, :].reshape(-1, 1) * b[i] for i in range(A.shape[0])]
+        return np.asarray(res)
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_empirical_risk/_linear_regression_cost.py]: I implemented this, I think its correct. Maybe you can take a look @nicola-bastianello, could you also check `LogisticRegressionCost`?

File: decent_bench/costs/_empirical_risk/_pytorch_cost.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    """
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        model: torch.nn.Module,
+        loss_fn: torch.nn.Module,
+        final_activation: torch.nn.Module | None = None,
+        *,
+        batch_size: EmpiricalRiskBatchSize = "all",
```
Comment (nicola-bastianello) [inline decent_bench/costs/_empirical_risk/_pytorch_cost.py]: maybe as a default we could have "batch"?

File: decent_bench/costs/_empirical_risk/_pytorch_cost.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+
+        Args:
+            dataset (Dataset):
+                Dataset partition containing features and targets.
+                Transformations should be applied beforehand such as converting to tensors.
+                See torch.utils.data.Dataset for details.
+            model (torch.nn.Module): PyTorch neural network model.
+            loss_fn: (torch.nn.Module): PyTorch loss function.
+            final_activation (torch.nn.Module | None): Optional final activation layer to apply after
+                model output when predicting targets. E.g., argmax if classification and model outputs logits.
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_empirical_risk/_pytorch_cost.py]: yes that is correct

File: decent_bench/costs/_empirical_risk/_pytorch_cost.py [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+            batch_size (EmpiricalRiskBatchSize): Size of mini-batches for stochastic methods, or "all" for full-batch.
+            device (SupportedDevices): Device to run computations on.
+            use_dataloader (bool): Whether to use DataLoader for batching.
+                Can be beneficial for large datasets to avoid loading all data into memory.
+            dataloader_kwargs (dict | None): Additional arguments for the DataLoader.
+            load_dataset (bool): If True, load the entire dataset into memory to optimize data access.
+                This may lead to major speedups if the dataset is lazily loaded.
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_empirical_risk/_pytorch_cost.py]: Is this better?
```
load_dataset (bool): If True, loads the entire dataset into memory to optimize data access.
                This may lead to major speedups if the dataset is lazily loaded (e.g., loading data from disk), but it
                might increase memory usage so set to False if memory is an issue.
```

It will lead to better performance if the dataset is lazily loaded since we load the data once from disk, not every time we access a datapoint

File: decent_bench/costs/_empirical_risk/_pytorch_cost.py [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+        return float("nan")
+
+    @property
+    def m_cvx(self) -> float:
+        return float("nan")
+
+    @cached_property
+    @override
+    def _rand(self) -> torch.Generator:  # type: ignore[override]
+        return torch.Generator(device="cpu").manual_seed(0)  # Later replace with global rng
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_empirical_risk/_pytorch_cost.py]: It should be done in issue #237. I can add a short comment to the issue as well

File: decent_bench/costs/_empirical_risk/_empirical_risk_cost.py [hunk 7]
```diff
@@ -118,6 +125,16 @@ def gradient(self, x: Array, indices: EmpiricalRiskIndices = "batch", **kwargs:
             - list[int]: corresponding datapoints are used.
             - "all": the full dataset is used.
             - "batch": a batch is drawn with :attr:`batch_size` samples.
+
+        Supported values for reduction are:
+            - "mean": average the gradients over the samples.
+            - None: return the gradients for each sample as a list.
+
+        Note:
+            When reduction is None, the returned array will have an additional leading dimension
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_empirical_risk/_empirical_risk_cost.py]: Github has been acting up all day for me, I am sorry if I've spammed you with comments. 

Eitherway, its not supposed to be a list. I accidentally wrote list on the argument documentation but the `Note` part is correct

File: decent_bench/metrics/metric_utils.py [hunk 8]
```diff
@@ -156,6 +188,210 @@ def iterative_convergence_rate_and_order(agent: AgentMetricsView, problem: Bench
...
+    Args:
+        agents: sequence of agents to calculate accuracy for
+        problem: benchmark problem containing test data
+        iteration: iteration to calculate accuracy at, or -1 to use the agents' final x
+
+    Returns:
+        list of accuracies per agent at *iteration*
+
+    """
+    if problem.test_data is None:
```
[the remark this answers is not in the document]
Comment (nicola-bastianello) [reply decent_bench/metrics/metric_utils.py]: you are right, overfitting could be a problem. let's keep the current implementation then
